dft for symmetric group algebra when p|n!

jacksonwalters opened the pull request:
The DFT for the SGA throws a ZeroDivisionError when p|n. Uses the idempotents as computed in [Murphy '83] to build the Pierce decomposition of the symmetric group algebra over a finite field. There is a homomorphic projection onto each block via v |--> v*e_i. The change of basis matrix is the modular Fourier transform, and works even when p|n.

<!-- ^ Please provide a concise and informative title. -->
<!-- ^ Don't put issue numbers in the title, do this in the PR description below. -->
<!-- ^ For example, instead of "Fixes #12345" use "Introduce new method to calculate 1 + 2". -->
<!-- v Describe your changes below in detail. -->
<!-- v Why is this change required? What problem does it solve? -->
<!-- v If this PR resolves an open issue, please link to it here. For example, "Fixes #12345". -->



### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### :hourglass: Dependencies

<!-- List all open PRs that this PR logically depends on. For example, -->
<!-- - #12345: short description why this is a dependency -->
<!-- - #34567: ... -->

jacksonwalters committed:
dft for symmetric group algebra when p|n

The DFT for the SGA throws a ZeroDivisionError p|n. Uses the idempotents as computed in [Murphy '83] to build the Pierce decomposition of the group algebra over a finite field. The "modular" Fourier transform is the homomorphic projection onto each block via v |--> v*e_i. This change of basis matrix is the modular Fourier transform, and works even when p|n.

jacksonwalters committed:
Update symmetric_group_algebra.py

jacksonwalters committed:
remove trailing whitespace

attempting to get through lint

jacksonwalters committed:
finish adapting from Jupyter notebook

jacksonwalters committed:
put all code into single lines

remove helper functions in favor of lambda functions

jacksonwalters committed:
Update symmetric_group_algebra.py

jacksonwalters committed:
Update symmetric_group_algebra.py

jacksonwalters committed:
try formatting matrix

jacksonwalters committed:
example was wrong

use p=2, n=3 for a small example.

jacksonwalters committed:
matrix is misspelled

lowercase m in matrix

jacksonwalters committed:
use transpose to get matrix of coordinate vectors

S17A05 commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -1909,12 +1909,14 @@ def dft(self, form="seminormal", mult='l2r'):
         """
         if form == "seminormal":
             return self._dft_seminormal(mult=mult)
```
It might be a good idea to catch the error in the case $p \mid n$ here and raise an error with a clearer error message, or to remove the standard argument for `form` and choose it depending on whether the characteristic divides $n$ (depending on whether `._dft_seminormal` always fails in the modular case)

jacksonwalters replied:
I agree, at the very least the failing cases p|n should at least be handled with a clear error message. I believe it only fails when p | n! (n was a type - it's the order of the symmetric group, so n!). I've chosen to just remove the standard argument for form and return the modular DFT in that case.

S17A05 commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -1941,6 +1943,42 @@ def _dft_seminormal(self, mult='l2r'):
+    def _dft_modular(self):
+        """
+        Return the discrete Foruier transform when the characterisc divides the order of the group.
+        The usual .dft() throws ZeroDivisionError when p|n.
```
I'd remove the reference to the `.dft()`-method here; there's also a small typo.
```suggestion
        Return the discrete Foruier transform when the characteristic divides the order of the group.
```

S17A05 commented on src/sage/combinat/symmetric_group_algebra.py:
```diff
@@ -1941,6 +1943,42 @@ def _dft_seminormal(self, mult='l2r'):
+            [0 0 1 0 0 1]
+            [0 0 0 1 1 0]
+            [1 0 0 1 1 0]
+            [0 1 1 0 0 1]
```
This example should go into the `.dft`-method, so that it shows up in the documentation, with the small change given below:

```suggestion
            sage: GF2S3 = SymmetricGroupAlgebra(GF(2),3)
            sage: GF2S3.dft(form="modular")
            [1 0 0 0 1 0]
            [0 1 0 0 0 1]
            [0 0 1 0 0 1]
            [0 0 0 1 1 0]
            [1 0 0 1 1 0]
            [0 1 1 0 0 1]
```

S17A05 reviewed (commented):
Admittedly I don't know enough about modular representation theory to check this code mathematically, but here are some programming comments.

Also, the documentation for the `.dft`-method should be rewritten a bit to explain the `form` argument, to make the explanation of `mult` dependent on `form = "seminormal"`, and to also explain the different approaches used depending on `form`. For the explanation of the modular approach, you can also refer to [Murphy '83], which needs to be formatted as `[Mur1983]_` in the docstring to give a hyperlink in the documentation.

jacksonwalters committed:
Update src/sage/combinat/symmetric_group_algebra.py

remove the self-reference to .dft(). just state that it's computing the DFT when the characteristic divides the order of the group.

Co-authored-by: Sebastian A. Spindler <153911337+S17A05@users.noreply.github.com>

jacksonwalters committed:
Update src/sage/combinat/symmetric_group_algebra.py

Co-authored-by: Sebastian A. Spindler <153911337+S17A05@users.noreply.github.com>

jacksonwalters committed:
Update symmetric_group_algebra.py

jacksonwalters committed:
add commentary explaining the DFT in this case

remove form from argument
